---
author: Margaret R. Starostik
title: "EMTAB3929: Data Substructure"
output:
  pdf_document:
    latex_engine: xelatex
mainfont: Arial
monofont: Arial
fontsize: 10pt
geometry: margin = 1in
header-includes:
  \usepackage[dvipsnames]{xcolor}
---

```{r GenSetup, include = FALSE}
# Ensure clean R environment
rm(list=ls())

# Change global default setting so every data frame created will not auto-convert to factors unless explicitly instructed
options(stringsAsFactors = FALSE) 

# Modify global options for compiling PDF
knitr::opts_chunk$set(tidy.opts = list(width.cutoff = 60), tidy = TRUE)

# Source my collection of functions
source("/Users/Margaret/Desktop/JHU/COURSEWORK/2018_Fall/Rotation02_RajivMcCoy/AneuploidyProject/Scripts/UsefulFunctions.R")

# Load/install packages
list_of_packages <- c("biomaRt", "BiocStyle", "cowplot", "devtools", "dplyr", "gplots", "gridExtra", "lme4", 
                      "MultiAssayExperiment", "readxl", "Rtsne", "scater", "scploid", "scran",
                      "tidyr", "umap")
install_and_load_packages(list_of_packages)

# Note that the edgeR package masks the scploid getCounts object. Resolve.
getCounts <- scploid::getCounts
```

## **1. Human Embryo EMTAB3929 scRNA-seq analysis**  
Data were acquired for this project as follows: 
(1) EMTAB3929 data (PMID 27062923) were downloaded on 11/6/2018 from http://imlspenticton.uzh.ch:3838/conquer/. Data included (a) MultiAssay Experiment (EMTAB3929.rds), (b) MultiQC report, (c) scater report, and (d) salmon archive (EMTAB3929_salmo.tar and EMTAB3929 folder).
(2) Supplementary files from Griffiths et al., 2017 were forked on 11/06/2018 from MarioniLab/Aneuploidy2017 on Github:  https://github.com/MarioniLab/Aneuploidy2017.
(3) Data from Griffiths et al., 2017 were downloaded on 11/14/2018 using the shell script supplied in their supplemnetary files (sh get_data.sh)

```{r ProjectSetup, include = FALSE}
# Set up folders
project_folder <- "/Users/Margaret/Desktop/JHU/COURSEWORK/2018_Fall/Rotation02_RajivMcCoy/AneuploidyProject/"

# Source functions from Griffiths' et al., 2017 and load data from mouse 8-cell stage G&T-seq and human Trisomy 21 G&T-seq
#griffiths_folder <- paste0(project_folder, "RawData/griffiths2017/")
#source(paste0(griffiths_folder, "aneu_functions.R"))

#load(file = paste0(griffiths_folder, "proc_data/emb8_data.RData"))
#load(file = paste0(griffiths_folder, "proc_data/tris_data.RData"))

# Load processed EMTAB3929 data and check data
load(paste0(project_folder, "ProcessedData/EMTAB3929_DataPrep.RData"))
#dim(annotation) # 56,400 genes
#dim(filtered_cpm) # 2,996 genes and 1,463 cells
#dim(log2_filtered_cpm) # 2,996 genes and 1,463 cells
#dim(metasheet) # 1,463 samples
```


## *1.1 Data Substructure* 
PCA was performed on log~2~(CPM + 1) gene expression values using the entire data set and separately for each embryonic stage.
```{r PCA, fig.show = "hold", fig.align = "center", fig.cap = "PCA", out.width = "40%", dev = "pdf"}
pca <- prcomp(t(log2_filtered_cpm))
variance <- pca$sdev^2

pca_data <- as.data.frame(pca$x[, 1:3])
pca_data$Sample <- rownames(pca_data)
pca_data <- inner_join(pca_data, metasheet, by = "Sample")

# PCA 
pca_plot1a <- ggplot(pca_data, aes(x = PC1, y = PC2, col = `Revised lineage (this study)`, shape = EStage)) + geom_point(alpha = 0.4)
pca_plot1b <- pca_plot1a + labs(x = paste0("PC1 ", format(variance[1] / sum(variance)*100, digits = 3), "%"), y = paste0("PC2 ", format(variance[2] / sum(variance)*100, digits = 3), "%"))
pca_plot1c <- pca_plot1b + theme_bw() + scale_color_brewer(palette = "Set1") + scale_shape_manual(values = c(15, 16, 17, 18, 3)) + labs(colour = "Cell Lineage", shape = "Stage")
#return(pca_plot1c)

pca_plot2a <- ggplot(pca_data, aes(x = PC1, y = PC3, col = `Revised lineage (this study)`, shape = EStage)) + geom_point(alpha = 0.4)
pca_plot2b <- pca_plot2a + labs(x = paste0("PC1 ", format(variance[1] / sum(variance)*100, digits = 3), "%"), y = paste0("PC3 ", format(variance[3] / sum(variance)*100, digits = 3), "%"))
pca_plot2c <- pca_plot2b + theme_bw() + scale_color_brewer(palette = "Set1") + scale_shape_manual(values = c(15, 16, 17, 18, 3)) 
#return(pca_plot2c)


plots <- plot_grid(pca_plot1c + theme(legend.position = "none"), pca_plot2c + theme(legend.position = "none"), align = "vh", nrow = 1, hjust = -1)
legend <- get_legend(pca_plot1c + guides(color = guide_legend(nrow = 1)) + theme(legend.position = "bottom"))

pdf(paste0(project_folder, "Results/PCA/EMTAB3929_PCA_AllSamples.pdf"), width = 12, height = 4)
plot_grid(plots, legend, ncol = 1, rel_heights = c(1, 0.1))
dev.off()
```

```{r StagePCA}
# Plot PCA separately for each of the stages
stages <- c("E3", "E4", "E5", "E6", "E7")
start_sample <- c(1, 79, 246, 606, 1011)
end_sample <- c(78, 245, 605, 1010, 1463)

compiled_spca <- vector("list", length(stages))
i <- 1
while (i <= length(stages)){
  spca <- t(log2_filtered_cpm)
  spca <- spca[start_sample[i]:end_sample[i], ]
  spca <- prcomp(spca)
  
  variance <- spca$sdev^2
  
  spca_data <- as.data.frame(spca$x[, 1:2])
  
  spca_plot1a <- ggplot(spca_data, aes(x = PC1, y = PC2)) + geom_point(color = "azure4")
  spca_plot1b <- spca_plot1a + labs(subtitle = paste0("Stage: ", stages[i], " (", dim(spca_data[1]), " cells)"),
                                  x = paste0("PC1 ", format(variance[1] / sum(variance)*100, digits = 3), "%"),
                                  y = paste0("PC2 ", format(variance[2] / sum(variance)*100, digits = 3), "%"))
  spca_plot1c <- spca_plot1b + theme_bw() + guides(fill = FALSE)
  
  compiled_spca[[i]] <- spca_plot1c
  
  i = i + 1
}

pdf(paste0(project_folder, "Results/PCA/EMTAB3929_PCA_ByStage.pdf"), width = 10, height = 5)
plot_grid(compiled_spca[[1]], compiled_spca[[2]], compiled_spca[[3]], compiled_spca[[4]], compiled_spca[[5]],
          align = "vh", nrow = 2, hjust = -1)
dev.off()
```

```{r tSNE}
# https://hms-dbmi.github.io/scw/heterogeneity.html
d <- dist(t(log2_filtered_cpm))

set.seed(7) # tSNE has some stochastic steps

tsne_5 <- Rtsne(d, is_distance = TRUE, perplexity = 5) # 1,000 iterations
tsne_30 <- Rtsne(d, is_distance = TRUE, perplexity = 30) # 1,000 iterations
tsne_50 <- Rtsne(d, is_distance = TRUE, perplexity = 50) # 1,000 iterations
tsne_100 <- Rtsne(d, is_distance = TRUE, perplexity = 100) # 1,000 iterations
tsne_200 <- Rtsne(d, is_distance = TRUE, perplexity = 200) # 1,000 iterations

tsne_data <- data.frame(
  Estage = metasheet$EStage,
  CellLineage = metasheet$`Revised lineage (this study)`,
  tsne5_1 = tsne_5$Y[, 1],
  tsne5_2 = tsne_5$Y[, 2],
  
  tsne30_1 = tsne_30$Y[, 1],
  tsne30_2 = tsne_30$Y[, 2],
  
  tsne50_1 = tsne_50$Y[, 1],
  tsne50_2 = tsne_50$Y[, 2],
  
  tsne100_1 = tsne_100$Y[, 1],
  tsne100_2 = tsne_100$Y[, 2],
  
  tsne200_1 = tsne_200$Y[, 1],
  tsne200_2 = tsne_200$Y[, 2])

compiled_tsne <- vector("list", 4)
i <- 3
while (i <= dim(tsne_data)[2]){
  
  tsne_subset <- tsne_data[, c(1, 2, i, i+1)]
  # Extract labels for x- and y-axis
  xname <- colnames(tsne_subset)[3]
  yname <- colnames(tsne_subset)[4]
  # Rename x- and y-axis for ggplot
  colnames(tsne_subset)[3] <- "x"
  colnames(tsne_subset)[4] <- "y"
  
  tsne_plot1a <- ggplot(tsne_subset, aes(x = x, y = y, color = Estage, shape = CellLineage)) + geom_point(alpha = 0.5, aes(size = ifelse(Estage %in% c("E3", "E4", "E5", "E6", "E7"), 1, NA)))
  
  tsne_plot1b <- tsne_plot1a + theme_bw() + scale_color_manual(values = c(rgb(133, 133, 169, maxColorValue = 255), rgb(82, 125, 157, maxColorValue = 255), rgb(120, 160, 90, maxColorValue = 255), rgb(219, 184, 57, maxColorValue = 255), rgb(183, 94, 81, maxColorValue = 255))) + scale_shape_manual(values = c(15, 16, 17, 18, 6, 8)) + guides(size = "none") + scale_size_continuous(range = c(1, 4)) + labs(x = "tSNE 1", y = "tSNE 2")
  
  compiled_tsne[[i]] <- tsne_plot1b
  
  i = i + 2
}

plots <- plot_grid(compiled_tsne[[3]] + theme(legend.position = "none") + labs(subtitle = "Complexity = 5"), 
                   compiled_tsne[[5]] + theme(legend.position = "none") + labs(subtitle = "Complexity = 30"), 
                   compiled_tsne[[7]] + theme(legend.position = "none") + labs(subtitle = "Complexity = 50"),
                   compiled_tsne[[9]] + theme(legend.position = "none") + labs(subtitle = "Complexity = 100"),
                   compiled_tsne[[11]] + theme(legend.position = "none") + labs(subtitle = "Complexity = 200"),
                   align = "vh", nrow = 1, hjust = -1)
legend <- get_legend(compiled_tsne[[3]] + guides(color = guide_legend(nrow = 1)) + theme(legend.position = "bottom"))

pdf(paste0(project_folder, "Results/tSNE/EMTAB3929_tSNE.pdf"), width = 14, height = 4)
plot_grid(plots, legend, ncol = 1, rel_heights = c(1, 0.15))
dev.off()
```

```{r tsne100_intermediates}
# Several intermediate plots are made for rotation talk presentation purposes to better depict results

# (1) Plot E3; rest are invisible so that the overall plot remains the same for all plots
tsne100_plot1a <- ggplot(tsne_data, aes(x = tsne100_1, y = tsne100_2, color = Estage, shape = CellLineage)) + geom_point(alpha = 0.5, aes(size = ifelse(Estage == "E3", 0.5, NA)))
tsne100_plot1b <- tsne100_plot1a + theme_bw() + scale_color_manual(values = c(rgb(133, 133, 169, maxColorValue = 255), rgb(82, 125, 157, maxColorValue = 255), rgb(120, 160, 90, maxColorValue = 255), rgb(219, 184, 57, maxColorValue = 255), rgb(183, 94, 81, maxColorValue = 255))) + scale_shape_manual(values = c(15, 16, 17, 18, 6, 8)) + guides(size = "none") + scale_size_continuous(range = c(1, 4)) + labs(x = "tSNE 1", y = "tSNE 2")

# (2) Plot E3 and E4 
tsne100_plot2a <- ggplot(tsne_data, aes(x = tsne100_1, y = tsne100_2, color = Estage, shape = CellLineage)) + geom_point(alpha = 0.5, aes(size = ifelse(Estage %in% c("E3", "E4"), 0.5, NA)))
tsne100_plot2b <- tsne100_plot2a + theme_bw() + scale_color_manual(values = c(rgb(133, 133, 169, maxColorValue = 255), rgb(82, 125, 157, maxColorValue = 255), rgb(120, 160, 90, maxColorValue = 255), rgb(219, 184, 57, maxColorValue = 255), rgb(183, 94, 81, maxColorValue = 255))) + scale_shape_manual(values = c(15, 16, 17, 18, 6, 8)) + guides(size = "none") + scale_size_continuous(range = c(1, 4)) + labs(x = "tSNE 1", y = "tSNE 2")

# (3) Plot E3, E4, and E5
tsne100_plot3a <- ggplot(tsne_data, aes(x = tsne100_1, y = tsne100_2, color = Estage, shape = CellLineage)) + geom_point(alpha = 0.5, aes(size = ifelse(Estage %in% c("E3", "E4", "E5"), 0.5, NA)))
tsne100_plot3b <- tsne100_plot3a + theme_bw() + scale_color_manual(values = c(rgb(133, 133, 169, maxColorValue = 255), rgb(82, 125, 157, maxColorValue = 255), rgb(120, 160, 90, maxColorValue = 255), rgb(219, 184, 57, maxColorValue = 255), rgb(183, 94, 81, maxColorValue = 255))) + scale_shape_manual(values = c(15, 16, 17, 18, 6, 8)) + guides(size = "none") + scale_size_continuous(range = c(1, 4)) + labs(x = "tSNE 1", y = "tSNE 2")

# (4) Plot E3, E4, E5, and E6
tsne100_plot4a <- ggplot(tsne_data, aes(x = tsne100_1, y = tsne100_2, color = Estage, shape = CellLineage)) + geom_point(alpha = 0.5, aes(size = ifelse(Estage %in% c("E3", "E4", "E5", "E6"), 0.5, NA)))
tsne100_plot4b <- tsne100_plot4a + theme_bw() + scale_color_manual(values = c(rgb(133, 133, 169, maxColorValue = 255), rgb(82, 125, 157, maxColorValue = 255), rgb(120, 160, 90, maxColorValue = 255), rgb(219, 184, 57, maxColorValue = 255), rgb(183, 94, 81, maxColorValue = 255))) + scale_shape_manual(values = c(15, 16, 17, 18, 6, 8)) + guides(size = "none") + scale_size_continuous(range = c(1, 4)) + labs(x = "tSNE 1", y = "tSNE 2")

# (4) Plot E3, E4, E5, E6, and E7
tsne100_plot5a <- ggplot(tsne_data, aes(x = tsne100_1, y = tsne100_2, color = Estage, shape = CellLineage)) + geom_point(alpha = 0.5, aes(size = ifelse(Estage %in% c("E3", "E4", "E5", "E6", "E7"), 0.5, NA)))
tsne100_plot5b <- tsne100_plot5a + theme_bw() + scale_color_manual(values = c(rgb(133, 133, 169, maxColorValue = 255), rgb(82, 125, 157, maxColorValue = 255), rgb(120, 160, 90, maxColorValue = 255), rgb(219, 184, 57, maxColorValue = 255), rgb(183, 94, 81, maxColorValue = 255))) + scale_shape_manual(values = c(15, 16, 17, 18, 6, 8)) + guides(size = "none") + scale_size_continuous(range = c(1, 4)) + labs(x = "tSNE 1", y = "tSNE 2")


plots <- plot_grid(tsne100_plot1b + theme(legend.position = "none") + labs(subtitle = "E3"), 
                   tsne100_plot2b + theme(legend.position = "none") + labs(subtitle = "E3 and E4"), 
                   tsne100_plot3b + theme(legend.position = "none") + labs(subtitle = "E3, E4, and E5"),
                   tsne100_plot4b + theme(legend.position = "none") + labs(subtitle = "E3, E4, E5, and E6"),
                   tsne100_plot5b + theme(legend.position = "none") + labs(subtitle = "E3, E4, E5, E6, and E7"),
                   align = "vh", nrow = 1, hjust = -1)
legend <- get_legend(tsne100_plot5b + guides(color = guide_legend(nrow = 1)) + theme(legend.position = "bottom"))

pdf(paste0(project_folder, "Results/tSNE/EMTAB3929_tSNE_Complexity100.pdf"), width = 15, height = 4)
plot_grid(plots, legend, ncol = 1, rel_heights = c(1, 0.15))
dev.off()
```
